Custom package order

isohedronpipeline opened the pull request.

isohedronpipeline committed:
Added CustomPackageOrder

maxnbk committed:
Tick off several completed boxes in onboarding document (#1691)

* Tick off several completed boxes in onboarding document

Signed-off-by: Stephen Mackenzie <maxnbk@users.noreply.github.com>

* Update ASWF/ONBOARDING.md

Signed-off-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

---------

Signed-off-by: Stephen Mackenzie <maxnbk@users.noreply.github.com>
Signed-off-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>
Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

isohedronpipeline committed:
Added PyPAPackageOrder

chadrik commented on src/rez/package_order.py:
```diff
@@ -608,6 +609,265 @@ def from_pod(cls, data):
 
+class CustomPackageOrder(PackageOrder):
+    """A package order that allows explicit specification of version ordering.
+    Specified through the "packages" attributes, which should be a dict which
```
some newlines in this docstring would make it more legible

chadrik commented on src/rez/package_order.py:
```diff
@@ -608,6 +609,265 @@ def from_pod(cls, data):
+        return priority_sort_key, version_key
+
+    @classmethod
+    def _packages_to_pod(cls, packages: Dict[str, List[Union[str, VersionRange]]]):
```
missing a return annotation

chadrik commented on src/rez/package_order.py:
```diff
@@ -608,6 +609,265 @@ def from_pod(cls, data):
+        }
+
+    @classmethod
+    def _packages_from_pod(cls, packages: Dict[str, List[Union[str, VersionRange]]]):
```
missing a return annotation

chadrik commented on src/rez/package_order.py:
```diff
@@ -608,6 +609,265 @@ def from_pod(cls, data):
+        elif not prerelease:
+            # We have no risk tolerance, so sort prereleases to the bottom.
+            risk_allowance_token = -pypa.Infinity
+        elif pre >= (prerelease, ):
```
is `pre` a tuple?   would be nice to have a type annotation for the variable above.

chadrik commented on src/rez/package_order.py:
```diff
@@ -608,6 +609,265 @@ def from_pod(cls, data):
+    """
+    name = "pypa"
+
+    def __init__(self, prerelease: Literal["", "rc", "b", "a"] = "", packages=None):
```
it could be a nice feature to allow "beta" and "alpha" as valid input options, and to map them to "b" and "a" for the value of `self.prerelease`. 

you an make a type variable to hold and reuse this literal at the top of the module:

```python
PrereleaseType: TypeAlias = Literal["rc", "b", "beta", "a", "alpha"]
```

chadrik replied:
From the pep440 spec: "Pre-releases allow the additional spellings of alpha, beta, c, pre, and preview for a, b, rc, rc, and rc respectively."

chadrik commented on src/rez/package_order.py:
```diff
@@ -608,6 +609,265 @@ def from_pod(cls, data):
+        super().__init__(list(packages))
+        self.packages_dict = self._packages_from_pod(packages)
+        if version_orderer and not isinstance(version_orderer, PackageOrder):
+            version_orderer = from_pod(version_orderer)
```
Can we fore `version_orderer` to always be of type `PackageOrder`?   And can we handle the `Dict[str, Any]` scenario in `from_pod`?  I've grown to prefer `__init__` methods to use the fully "structured" typed, and use classmethods to handle converting "unstructured" types to "structured" types.

isohedronpipeline replied:
Yeah, should be easy enough.

chadrik commented on src/rez/package_order.py:
```diff
@@ -608,6 +609,265 @@ def from_pod(cls, data):
+                for that package
+            version_orderer (Optional[Union[PackageOrder, Dict[str, Any]]]):
+                How versions are sorted within version ranges.
+                Can take a pod representation of an orderer or a PackageOrder object
```
this seems like a nice solution to the chaining problem.    Any downsides that we should document?

isohedronpipeline replied:
No, it's just specific to the custom package order.  Would be nice to be able to arbitrarily chain them together in a general way.

chadrik reviewed (commented):
nice work.

loonghao committed:
Add retry mechanism when renaming folders on Windows (#1666)

This will fix issues when rez-cp could fail in some situations on WIndows.

Signed-off-by: loonghao <hal.long@outlook.com>

JeanChristopheMorinPerso committed:
Add -e/--editable flag to install script to create an editable install (#1672)

Signed-off-by: Jean-Christophe Morin <jean_christophe_morin@hotmail.com>

isohedronpipeline committed:
Apply package orderers on variants (#1684)

Previously, package orderers were only applied on packages. This was an oversight. With this change, they are now applied to variants too.

---------

Signed-off-by: Ben Andersen <ben@isohedron.com.au>
Co-authored-by: Paul Molodowitch <elrond79@gmail.com>

BryceGattis committed:
Add releasing packages documentation (#1689)

Signed-off-by: Bryce Gattis <brycegattis@yahoo.com>

BryceGattis committed:
Improve clarify of early binding functions documentation (#1677)

Improve clarify of early binding functions documentation by documenting what is accessible within an early function.


Signed-off-by: Bryce Gattis <brycegattis@yahoo.com>

JeanChristopheMorinPerso committed:
Bump version to 3.1.0 (#1702)

Signed-off-by: Jean-Christophe Morin <jean_christophe_morin@hotmail.com>
